libr/bin/format/elf/elf.c

```diff
@@ -1277,454 +1296,499 @@ ut64 Elf_(r_bin_elf_get_section_addr_end)(ELFOBJ *bin, const char *section_name)
 	}
 
-	if (bin->is_rela == DT_REL) {
-		rel_sec = get_rel_sec (bin, rel_sect);
-		tsize = sizeof (Elf_(Rel));
-	} else if (bin->is_rela == DT_RELA) {
-		rel_sec = get_rel_sec (bin, rela_sect);
-		tsize = sizeof (Elf_(Rela));
-	}
-	if (!rel_sec || rel_sec->size < 1) {
-		return NULL;
-	}
+	size_t i = 0;
+	Elf_(Rela) reloc_info;
 
-	nrel = (ut32) ((int)rel_sec->size / (int)tsize);
-	if (nrel < 1) {
-		return NULL;
-	}
+	reloc_info.r_offset = READWORD (buf, i);
+	reloc_info.r_info = READWORD (buf, i);
 
-	const int htsize = R_MIN (nrel, 1024);
-	rel_cache = ht_up_new_size (htsize, NULL, rel_cache_free, NULL);
+	if (rel_mode == DT_RELA) {
+		reloc_info.r_addend = READWORD (buf, i);
+		r->addend = reloc_info.r_addend;
+	}
 
-	for (j = k = 0; j < rel_sec->size && k < nrel; j += tsize, k++) {
-		int len = r_buf_read_at (bin->b, rel_sec->offset + j, rl, tsize);
-		if (len != tsize) {
-			break;
-		}
-		struct ht_rel_t *rel = read_ht_rel (bin, rl, k);
-		if (!rel) {
-			goto out;
-		}
+	r->rel_mode = rel_mode;
+	r->last = 0;
+	r->offset = reloc_info.r_offset;
+	r->sym = ELF_R_SYM (reloc_info.r_info);
+	r->type = ELF_R_TYPE (reloc_info.r_info);
 
-		if (!ht_up_insert (rel_cache, REL_SYM, rel)) {
-			free (rel);
-		}
-	}
-	return rel_cache;
-out:
-	ht_up_free (rel_cache);
-	return NULL;
+	return true;
 }
 
-static ut64 get_dyn_entry(ELFOBJ *bin, int dyn_entry) {
-	int i;
-	for (i = 0; i < bin->dyn_entries; i++) {
-		if (bin->dyn_buf[i].d_tag == dyn_entry) {
-			switch (bin->dyn_buf[i].d_tag) {
-			case DT_REL:
-			case DT_RELA:
-			case DT_PLTGOT:
-			case DT_JMPREL:
-				return bin->dyn_buf[i].d_un.d_ptr;
-			case DT_RELSZ:
-			case DT_RELASZ:
-			case DT_PLTRELSZ:
-				return bin->dyn_buf[i].d_un.d_val;
-			default:
-				r_warn_if_reached ();
-				break;
-			}
-		}
+static size_t get_num_relocs_dynamic_plt(ELFOBJ *bin) {
+	if (bin->dyn_info.dt_pltrelsz) {
+		return bin->dyn_info.dt_pltrelsz / get_size_rel_mode (bin->dyn_info.dt_pltrel);
 	}
-
-	return -1;
+	return 0;
 }
 
-static ut64 get_got_addr(ELFOBJ *bin) {
-	ut64 ret = get_dyn_entry (bin, DT_PLTGOT);
-	if (ret != -1) {
-		return ret;
-	}
+static size_t get_num_relocs_dynamic(ELFOBJ *bin) {
+	size_t res = 0;
 
-	ret = Elf_(r_bin_elf_get_section_addr) (bin, ".got");
-	if (ret) {
-		return ret;
+	if (bin->dyn_info.dt_relaent) {
+		res += bin->dyn_info.dt_relasz / bin->dyn_info.dt_relaent;
 	}
 
-	ret = Elf_(r_bin_elf_get_section_addr) (bin, ".got.plt");
-	if (ret) {
-		return ret;
+	if (bin->dyn_info.dt_relent) {
+		res += bin->dyn_info.dt_relsz / bin->dyn_info.dt_relent;
 	}
 
-	return -1;
+	return res + get_num_relocs_dynamic_plt (bin);
 }
 
-static ut64 get_import_addr_ppc(ELFOBJ *bin, struct ht_rel_t *rel, RBinElfSection *plt_section, int nrel) {
-	ut8 buf[4] = { 0 };
-	if (!plt_section) {
-		return -1;
-	}
-	int len = r_buf_read_at (bin->b, plt_section->offset, buf, sizeof (buf));
-	if (len < 4) {
-		return -1;
-	}
+static bool sectionIsInvalid(ELFOBJ *bin, RBinElfSection *sect) {
```

Comment:
it's usually better, AFAIK, to use positive checks. So `sectionIsValid` seems better and then if you need the opposite you just use `!sectionIsValid`, which "reads" better IMO.